### PR TITLE
Better experience for disabled progress

### PR DESCRIPTION
**Pull request by Bjvanminnen**
### Old model:
On script overview
- Load the page along with any progress stored in sessionStorage
- Make an API call for user progress. Merge progress here with locally stored progress.
- If API call tells us we're signed in (and milestone posts are disabled) gray out bubbles. This means they'd have been visible with progress for a time.

On puzzle page:
- Load the page along with any progress stored in sessionStorage
- Make an API call for user progress. Merge progress here with locally stored progress.
- If API call tells us we're signed in (and milestone posts are disabled) gray out bubbles. This means they'd have been visible with progress for a time.

### New model:
On script overview
- Load the page along with any progress stored in sessionStorage. If postMilestone is disabled, don't show that progress until we know sign in state (which we'll find out via an API call).
- If API call tells us we're signed in (and milestone posts are disabled) gray out the bubbles, otherwise show progress

On puzzle page:
- Load the page along with any progress stored in sessionStorage. If postMilestone is disabled, don't show that progress until we know sign in state.
- For non-cached scripts, we'll know sign in state immediately. For cached scripts, we won't until we make an API call. However, we will cache this state in sessionStorage, such that after first page load we should know whether or not we're signed in without an additional API call.
- If API call tells us we're signed in (and milestone posts are disabled) gray out the bubbles, otherwise show progress

Summary:
The main difference then between the old and the new is that instead of showing progress, and then hiding it later, we will refrain from showing progress until we know if you're signed in or not (in the case that postMilestone is disabled). In many cases this will be immediately, but in some cases it will not be until after an API call.

This means that for signed out users, whereas in the past they would always immediately see their locally cached progress, they might now see it only after we make an API call.

### CODE REVIEW

File: apps/src/code-studio/header.js [hunk 1]
```diff
@@ -37,11 +41,19 @@ const PUZZLE_PAGE_NONE = -1;
...
  *   }>
  * }}
+ * @param {object} progressData
+ * @param {string} currentLevelId
+ * @param puzzlePage
+ * @param {boolean} [signedIn] True/false if we know the sign in state of the
+ *   user, null otherwise
  */
-header.build = function (stageData, progressData, currentLevelId, scriptName, puzzlePage) {
+header.build = function (scriptData, stageData, progressData, currentLevelId, puzzlePage, signedIn) {
```
Comment (Bjvanminnen) [inline apps/src/code-studio/header.js:50]: Moved scriptName into scriptData

File: apps/src/code-studio/progressRedux.js [hunk 2]
```diff
@@ -89,10 +96,25 @@ export default function reducer(state = initialState, action) {
...
 
-  if (action.type === DISABLE_BUBBLE_COLORS) {
+  if (action.type === DISABLE_POST_MILESTONE) {
+    return {
+      ...state,
+      postMilestoneDisabled: true
+    };
+  }
+
+  // TODO (brent) - write tests for new reducers
```
Comment (Bjvanminnen) [inline apps/src/code-studio/progressRedux.js:106]: Plan to address this before merging

File: apps/src/code-studio/header.js [hunk 3]
```diff
@@ -37,11 +41,19 @@ const PUZZLE_PAGE_NONE = -1;
  *     kind: string
  *   }>
  * }}
+ * @param {object} progressData
+ * @param {string} currentLevelId
+ * @param puzzlePage
```
Comment (islemaster) [inline apps/src/code-studio/header.js]: type?

File: apps/src/code-studio/progress.js [hunk 4]
```diff
@@ -141,9 +196,14 @@ progress.renderCourseProgress = function (scriptData, currentLevelId) {
...
- * @param currentLevelId
- * @param saveAnswersBeforeNavigation
+ * @param {object} scriptData
+ * @param {string} scriptData.name
+ * @param {boolean} scriptData.disablePostMilestone
+ * @param {boolean} [scriptData.plc]
+ * @param {object[]} [scriptData.stages]
+ * @param {boolean} [scriptData.peerReviewsRequired]
+ * @param {string} currentLevelId
+ * @param {boolean} [saveAnswersBeforeNavigation]
```
Comment (islemaster) [inline apps/src/code-studio/progress.js:206]: Maybe a `@typedef ScriptData` block would come in handy?

File: apps/src/code-studio/initApp/loadApp.js [hunk 5]
```diff
@@ -331,11 +316,19 @@ function loadAppAsync(appOptions) {
...
         }
 
-        const signedOutUser = Object.keys(data).length === 0;
-        if (!signedOutUser && (data.disablePostMilestone ||
-                               experiments.isEnabled('postMilestoneDisabledUI'))) {
-          getStore().dispatch(disableBubbleColors());
-          showDisabledButtonsAlert(!!data.isHoc);
+        const store = getStore();
+        const signInState = store.getState().progress.signInState;
+        if (signInState === SignInState.Unknown) {
```
Comment (islemaster) [inline apps/src/code-studio/initApp/loadApp.js:321]: Should this be checking whether `signInState` has changed, not whether it's unknown?  If we're caching this locally, what happens to a user that signs in, plays a few levels, then signs out again - does the signed-in state stay cached?

Comment (Bjvanminnen) [reply apps/src/code-studio/initApp/loadApp.js:321]: When you sign out, we clear your clientState.

File: apps/src/code-studio/progress.js (hunk 4 above)
Comment (Bjvanminnen) [reply apps/src/code-studio/progress.js:206]: I had that at first, but although we have a parameter called scriptData in a couple of places, it's contents are not always the same (though it always contains data about the script). I decided it was better to just document the options being passed in in each case.

**Pull request merged**